core/src/main/java/com/linecorp/armeria/common/HttpRequestBuilder.java

```diff
@@ -0,0 +1,397 @@
+     *            .queryParams(QueryParams.of("from", "foo", "limit", 10))
+     *            .build(); // GET `/endpoint?from=foo&limit=10`
+     * }</pre>
+     * @see QueryParams
+     */
+    public HttpRequestBuilder queryParams(QueryParams queryParams) {
+        requireNonNull(queryParams, "queryParams");
+        queryParams.forEach((BiConsumer<String, String>) queryParamsBuilder::set);
+        return this;
+    }
+
+    /**
+     * Sets a cookie for this request. For example:
+     * <pre>{@code
+     * HttpRequest.builder()
+     *            .get("/")
+     *            .cookie(Cookie.of("cookie", "foo"))
+     *            .build();
+     * }</pre>
+     * @see Cookie
+     */
+    public HttpRequestBuilder cookie(Cookie cookie) {
+        cookies.add(requireNonNull(cookie, "cookie"));
+        return this;
+    }
+
+    /**
+     * Sets multiple cookies for this request. For example:
+     * <pre>{@code
+     * HttpRequest.builder()
+     *            .get("/")
+     *            .cookies(Cookies.of(Cookie.of("cookie1", "foo"),
+     *                                Cookie.of("cookie2", "bar")))
+     *            .build();
+     * }</pre>
+     * @see Cookies
+     */
+    public HttpRequestBuilder cookies(Cookies cookies) {
+        this.cookies.addAll(requireNonNull(cookies, "cookies"));
+        return this;
+    }
+
+    /**
+     * Creates a new {@link HttpRequest}.
+     */
+    public HttpRequest build() {
+        final RequestHeaders requestHeaders = requestHeaders();
+        if (content == null || content.isEmpty()) {
+            if (content != null) {
+                content.close();
+            }
+            if (httpTrailersBuilder.isEmpty()) {
+                return new EmptyFixedHttpRequest(requestHeaders);
+            }
+            return new OneElementFixedHttpRequest(requestHeaders, httpTrailersBuilder.build());
+        }
+        if (httpTrailersBuilder.isEmpty()) {
+            return new OneElementFixedHttpRequest(requestHeaders, content);
+        }
+        return new TwoElementFixedHttpRequest(requestHeaders, content, httpTrailersBuilder.build());
+    }
+
+    private RequestHeaders requestHeaders() {
+        requestHeadersBuilder.path(buildPath());
+        if (!cookies.isEmpty()) {
+            requestHeadersBuilder.set(COOKIE, Cookie.toCookieHeader(cookies));
+        }
+        if (content == null || content.isEmpty()) {
+            requestHeadersBuilder.remove(CONTENT_LENGTH);
+        } else {
+            requestHeadersBuilder.setInt(CONTENT_LENGTH, content.length());
+        }
+        return requestHeadersBuilder.build();
+    }
+
+    private String buildPath() {
+        checkState(path != null, "path must be set.");
+        final StringBuilder pathBuilder = new StringBuilder(path);
+        if (!pathParams.isEmpty()) {
+            int i = 0;
+            while (i < pathBuilder.length()) {
+                if (pathBuilder.charAt(i) == '{') {
+                    int j = i;
```

Comment:
Maybe?
```suggestion
                    int j = i + 1;
```

And we need to check whether `j` is greater than `pathBuilder.length()` just in case a path ends with `/foo/bar{`

Reply:
I added this test. Let me know if there is anything missing.
https://github.com/line/armeria/blob/5e7c506c82aacd19fbb11a3268cc9bbe1e90d0e7/core/src/test/java/com/linecorp/armeria/common/HttpRequestBuilderTest.java#L114-L129